File: integration/unity/Assets/Scripts/behaviac/runtime/FSM/WaitState.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

namespace behaviac
{
    public class WaitState : State
    {
        protected override void load(int version, string agentType, List<property_t> properties)
        {
            base.load(version, agentType, properties);

            for (int i = 0; i < properties.Count; ++i)
            {
                property_t p = properties[i];

                if (p.name == "Time")
                {
                    int pParenthesis = p.value.IndexOf('(');

                    if (pParenthesis == -1)
                    {
                        this.m_time = AgentMeta.ParseProperty(p.value);
                    }
                    else
                    {
                        this.m_time = AgentMeta.ParseMethod(p.value);
                    }
                }
            }
        }

        protected virtual double GetTime(Agent pAgent)
        {
            double time = 0;

            if (this.m_time != null)
            {
                if (this.m_time is CInstanceMember<double>)
                {
                    time = ((CInstanceMember<double>)this.m_time).GetValue(pAgent);
                }
                else if (this.m_time is CInstanceMember<float>)
                {
                    time = ((CInstanceMember<float>)this.m_time).GetValue(pAgent);
                }
                else if (this.m_time is CInstanceMember<int>)
                {
                    time = ((CInstanceMember<int>)this.m_time).GetValue(pAgent);
                }
            }

            return time;
        }

        protected virtual int GetIntTime(Agent pAgent)
        {
            int time = 0;

            if (this.m_time != null)
            {
                if (this.m_time is CInstanceMember<int>)
                {
                    time = ((CInstanceMember<int>)this.m_time).GetValue(pAgent);
                }
            }

            return time;
        }

        protected IInstanceMember m_time;

        protected override BehaviorTask createTask()
        {
            WaitStateTask pTask = new WaitStateTask();

            return pTask;
        }

        private class WaitStateTask : State.StateTask
        {
            private double m_start = 0;
            private double m_time = 0;
            private long m_intStart = 0;
            private int m_intTime = 0;

            public override void copyto(BehaviorTask target)
            {
                base.copyto(target);

                Debug.Check(target is WaitStateTask);
                WaitStateTask ttask = (WaitStateTask)target;

                ttask.m_start = this.m_start;
                ttask.m_time = this.m_time;

                ttask.m_intStart = this.m_intStart;
                ttask.m_intTime = this.m_intTime;
            }

            public override void save(ISerializableNode node)
            {
                base.save(node);

                CSerializationID startId = new CSerializationID("start");
                node.setAttr(startId, this.m_start);

                CSerializationID timeId = new CSerializationID("time");
                node.setAttr(timeId, this.m_time);

                CSerializationID intStartId = new CSerializationID("intstart");
                node.setAttr(intStartId, this.m_intStart);

                CSerializationID intTimeId = new CSerializationID("inttime");
                node.setAttr(intTimeId, this.m_intTime);
            }

            public override void load(ISerializableNode node)
            {
                base.load(node);
            }

            private double GetTime(Agent pAgent)
            {
                WaitState pWaitNode = this.GetNode() as WaitState;

                return pWaitNode != null ? pWaitNode.GetTime(pAgent) : 0;
            }

            private int GetIntTime(Agent pAgent)
            {
                WaitState pWaitNode = this.GetNode() as WaitState;

                return pWaitNode != null ? pWaitNode.GetIntTime(pAgent) : 0;
            }

            protected override bool onenter(Agent pAgent)
            {
                this.m_nextStateId = -1;

                if (Workspace.Instance.UseIntValue)
                {
                    this.m_intStart = Workspace.Instance.IntValueSinceStartup;
                    this.m_intTime = this.GetIntTime(pAgent);

                    return (this.m_intTime >= 0);
                }
                else
                {
                    this.m_start = Workspace.Instance.DoubleValueSinceStartup;
                    this.m_time = this.GetTime(pAgent);

                    return (this.m_time >= 0);
                }
            }

            protected override void onexit(Agent pAgent, EBTStatus s)
            {
            }

            protected override EBTStatus update(Agent pAgent, EBTStatus childStatus)
            {
                Debug.Check(childStatus == EBTStatus.BT_RUNNING);
                Debug.Check(this.m_node is WaitState, "node is not an WaitState");
                WaitState pStateNode = (WaitState)this.m_node;

                if (Workspace.Instance.UseIntValue)
                {
                    if (Workspace.Instance.IntValueSinceStartup - this.m_intStart >= this.m_intTime)
                    {
                        pStateNode.Update(pAgent, out this.m_nextStateId);
                        return EBTStatus.BT_SUCCESS;
                    }
                }
                else
                {
                    if (Workspace.Instance.DoubleValueSinceStartup - this.m_start >= this.m_time)
                    {
                        pStateNode.Update(pAgent, out this.m_nextStateId);
                        return EBTStatus.BT_SUCCESS;
                    }
                }

                return EBTStatus.BT_RUNNING;
            }
        }
    }
}

File: src/fsm/waitstate.cpp
#include "behaviac/fsm/waitstate.h"
#include "behaviac/behaviortree/attachments/effector.h"
#include "behaviac/fsm/startcondition.h"
#include "behaviac/fsm/transitioncondition.h"
#include "behaviac/agent/agent.h"
#include "behaviac/behaviortree/nodes/conditions/condition.h"
#include "behaviac/common/meta.h"

namespace behaviac {
    WaitState::WaitState() : m_time(0) {
    }

    WaitState::~WaitState() {
        BEHAVIAC_DELETE(this->m_time);
    }

    void WaitState::load(int version, const char* agentType, const properties_t& properties) {
        super::load(version, agentType, properties);

        for (propertie_const_iterator_t it = properties.begin(); it != properties.end(); ++it) {
            const property_t& p = (*it);

            if (StringUtils::StringEqual(p.name, "Time")) {
                if (StringUtils::IsValidString(p.value)) {
                    const char* pParenthesis = StringUtils::StringFind(p.value, '(');

                    if (pParenthesis == 0) {
                        behaviac::string typeName;
                        this->m_time = AgentMeta::ParseProperty(p.value);
                    } else {
                        this->m_time = AgentMeta::ParseMethod(p.value);
                    }
                }
            }
        }
    }

    double WaitState::GetTime(Agent* pAgent) const {
        double time = 0;

        if (this->m_time != NULL) {
            int typeNumberId = this->m_time->GetClassTypeNumberId();

            if (typeNumberId == GetClassTypeNumberId<int>()) {
                time = *(int*)this->m_time->GetValue(pAgent);
            } else if (typeNumberId == GetClassTypeNumberId<double>()) {
                time = *(double*)this->m_time->GetValue(pAgent);
            } else if (typeNumberId == GetClassTypeNumberId<float>()) {
                time = *(float*)this->m_time->GetValue(pAgent);
            } else {
                BEHAVIAC_ASSERT(false);
            }
        }

        return time;
    }

    int WaitState::GetIntTime(Agent* pAgent) const {
        int time = 0;

		if (this->m_time) {
			int typeNumberId = this->m_time->GetClassTypeNumberId();

			if (typeNumberId == GetClassTypeNumberId<int>()) {
				time = *(int*)this->m_time->GetValue(pAgent);
			}
			else {
				BEHAVIAC_ASSERT(false);
			}
		}

        return time;
    }

    bool WaitState::IsValid(Agent* pAgent, BehaviorTask* pTask) const {
        if (WaitState::DynamicCast(pTask->GetNode()) == 0) {
            return false;
        }

        return super::IsValid(pAgent, pTask);
    }

    BehaviorTask* WaitState::createTask() const {
        WaitStateTask* pTask = BEHAVIAC_NEW WaitStateTask();

        return pTask;
    }

    WaitStateTask::WaitStateTask() : StateTask(), m_start(0), m_time(0), m_intStart(0), m_intTime(0) {
    }

    WaitStateTask::~WaitStateTask() {
    }

    void WaitStateTask::copyto(BehaviorTask* target) const {
        super::copyto(target);

        BEHAVIAC_ASSERT(WaitStateTask::DynamicCast(target));
        WaitStateTask* ttask = (WaitStateTask*)target;

        ttask->m_start = this->m_start;
        ttask->m_time = this->m_time;

        ttask->m_intStart = this->m_intStart;
        ttask->m_intTime = this->m_intTime;
    }

    void WaitStateTask::save(IIONode* node) const {
        super::save(node);

        if (this->m_status != BT_INVALID) {
            CIOID  startId("start");
            node->setAttr(startId, this->m_start);

            CIOID  timeId("time");
            node->setAttr(timeId, this->m_time);

            CIOID  intStartId("intstart");
            node->setAttr(intStartId, this->m_intStart);

            CIOID  intTimeId("inttime");
            node->setAttr(intTimeId, this->m_intTime);
        }
    }

    void WaitStateTask::load(IIONode* node) {
        super::load(node);

        if (this->m_status != BT_INVALID) {
            behaviac::string attrStr;

            CIOID  startId("start");
            node->getAttr(startId, attrStr);
            StringUtils::ParseString(attrStr.c_str(), this->m_start);

            CIOID  timeId("time");
            node->getAttr(timeId, attrStr);
            StringUtils::ParseString(attrStr.c_str(), this->m_time);

            CIOID  intStartId("intstart");
            node->getAttr(intStartId, attrStr);
            StringUtils::ParseString(attrStr.c_str(), this->m_intStart);

            CIOID  intTimeId("inttime");
            node->getAttr(intTimeId, attrStr);
            StringUtils::ParseString(attrStr.c_str(), this->m_intTime);
        }
    }

    double WaitStateTask::GetTime(Agent* pAgent) const {
        const WaitState* pWaitNode = WaitState::DynamicCast(this->GetNode());

        return pWaitNode ? pWaitNode->GetTime(pAgent) : 0;
    }

    int WaitStateTask::GetIntTime(Agent* pAgent) const {
        const WaitState* pWaitNode = WaitState::DynamicCast(this->GetNode());

        return pWaitNode ? pWaitNode->GetIntTime(pAgent) : 0;
    }

    bool WaitStateTask::onenter(Agent* pAgent) {
        BEHAVIAC_UNUSED_VAR(pAgent);

        this->m_nextStateId = -1;

        bool bUseIntValue = Workspace::GetInstance()->GetUseIntValue();

        if (bUseIntValue) {
            this->m_intStart = Workspace::GetInstance()->GetIntValueSinceStartup();
            this->m_intTime = this->GetIntTime(pAgent);

            if (this->m_intTime <= 0) {
                return false;
            }
        } else {
            this->m_start = Workspace::GetInstance()->GetDoubleValueSinceStartup();
            this->m_time = this->GetTime(pAgent);

            if (this->m_time <= 0) {
                return false;
            }
        }

        return true;
    }

    void WaitStateTask::onexit(Agent* pAgent, EBTStatus s) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(s);
    }

    EBTStatus WaitStateTask::update(Agent* pAgent, EBTStatus childStatus) {
        BEHAVIAC_ASSERT(childStatus == BT_RUNNING);
        BEHAVIAC_UNUSED_VAR(childStatus);
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_ASSERT(WaitState::DynamicCast(this->GetNode()) != 0, "node is not an WaitState");

        WaitState* pStateNode = (WaitState*)(this->GetNode());

		bool bUseIntValue = Workspace::GetInstance()->GetUseIntValue();

		if (bUseIntValue) {
			long long time = Workspace::GetInstance()->GetIntValueSinceStartup();

			if (time - this->m_intStart >= this->m_intTime) {
				pStateNode->Update(pAgent, this->m_nextStateId);
				return BT_SUCCESS;
			}
		}
		else {
			double time = Workspace::GetInstance()->GetDoubleValueSinceStartup();

			if (time - this->m_start >= this->m_time) {
				pStateNode->Update(pAgent, this->m_nextStateId);
				return BT_SUCCESS;
			}
		}

        return BT_RUNNING;
    }
}
